synonymdev/bitkit-react-native
# feat(wallet): Detect Reorgs And RBFs

### Description

- Bitkit now tracks transactions with less than 6-confirmations and notifies users of a relevant reorg or RBF.
- Added `unconfirmedTransactions` to `wallet` reducer.
- Created several new methods for tracking unconfirmed txs.
- Added a button to inject a fake transaction in the Dev menu.
- Added `exists` to `IActivityItem` types.
- Updated `mergeActivityItems` method.
- Added methods to clear store objects when rescanning.

With this update we're concerned with two main things that can happen to transactions with less than 6 confirmations.
 1. A transaction can get pushed back to the mempool due to a reorg.
 2. A zero-conf transaction can get removed from the mempool if it has insufficient fees for the given fee environment, a reorg occurs that knocks it from the mempool or it's RBF'd.

In the event number one occurs, Bitkit will detect that a previously confirmed transaction is now no longer confirmed and notifies the user accordingly. This is a best-case scenario where a user may only need to wait for the tx to confirm again. The only action that needs to be taken is ensuring the activity item is no longer marked as confirmed and the stored transaction height is adjusted.

In the event number two occurs, Bitkit will detect that a tracked transaction is no longer in the mempool and alert the user accordingly of a potential reorg or RBF. This situation is more severe since the transaction in question is now gone, reverting ownership of any received or sent funds. In this situation, Bitkit will set the `exists` state to false in the activity list item and highlight the tx entry in red in the activity list. Bitkit will also perform a rescan of the wallet to ensure funds and stored transactions are accurate.

### Type of change


### Tests


### Screenshot / Video
![ghost-tx](https://github.com/synonymdev/bitkit/assets/8532651/224ad542-594f-4599-9f74-5c5036ea2ee1)
![reorg](https://github.com/synonymdev/bitkit/assets/8532651/faf92e5c-a585-410c-adc2-813ff646079e)
![removed-from-mempool](https://github.com/synonymdev/bitkit/assets/8532651/ecffea60-dbef-4298-841f-c8f5a58d54a6)

### QA Notes

To test a reorg:
 1. Send funds to the wallet: `/regtest-deposit <address> 0.00005000`
 2. Mine a block: `/regtest-mine 1`
 3. Invalidate the block you just mined: `/regtest-invalidate-block <blockhash>`
 Expected: A blue notification should appear that matches the attached screenshot after detecting the reorg. The transaction should no longer appear as confirmed in the activity list.

To test an RBF or getting bumped from the mempool:
 1. Navigate to "Settings->Dev Settings->Inject Fake Transaction"
Expected: A red notification should appear that matches the attached screenshot after detecting the ghost/invalid tx. The tx in question should be red in the activity list. When tapping the activity list item the status should display "Removed from Mempool".

### Notes

- I added translations for the new notifications using Google translate. Please let me know if they need further adjustments.
